## 0% of end-end tests have passed
```
Total executed tests: 6
Failed tests: 6
Expected to fail tests: 0
Passed tests: 0
```
```
IDs of failed tests: 727230,702786,703629,702838,702775,704613 
```
<h3>Failed tests (6)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=feature/keycard-sign-tx&PR_ID=20845&APK_NAME=20845.apk&TR_CASE_IDS=727230,702786,703629,702838,702775,704613">Rerun failed tests</a></li><br/><h4>Class TestWalletMultipleDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3627172">test_wallet_send_asset_from_drawer</a>, id: 727230 </b></td></tr><tr><td><p><blockquote><div>Device 2: Find `Text` by `xpath`: `//android.view.ViewGroup[@content-desc='container']/android.widget.TextView[@text='Ether']/../android.widget.TextView[3]` </div><div>Device 2: `Text` is `0.02109 ETH` </div></blockquote></p>

```
critical/test_wallet.py:184: in test_wallet_send_asset_from_drawer
    self.errors.verify_no_errors()
base_test_case.py:191: in verify_no_errors
    pytest.fail('\n '.join([self.errors.pop(0) for _ in range(len(self.errors))]))
 Eth amount in the receivers wallet is 0.0211 but should be 0.0417
```

<br/><br/></td></tr></tbody></table><h4>Class TestCommunityMultipleDeviceMergedTwo:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3627168">test_community_mentions_push_notification</a>, id: 702786 </b></td></tr><tr><td><p><blockquote><div>Device 2: Find `LogInButton` by `accessibility id`: `login-button` </div><div>Device 2: Tap on found: LogInButton </div></blockquote></p>

```
Test setup failed: critical/chats/test_public_chat_browsing.py:868: in prepare_devices
    self.community_2.join_community()
../views/chat_view.py:425: in join_community
    self.community_status_joined.wait_for_visibility_of_element(60)
../views/base_element.py:147: in wait_for_visibility_of_element
    raise TimeoutException(
 Device 2: Text by accessibility id:`status-tag-positive` is not found on the screen after wait_for_visibility_of_element
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/d79d762dfc0d4e579faf3fad08cd4638?auth=601d87eb308cdcf53e6046297a8d70de">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/4d7d372d1d4d4f03bdbb865e3803332a?auth=84eb74440359f48405d8173b5abf927a">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3627170">test_community_join_when_node_owner_offline</a>, id: 703629 </b></td></tr><tr><td>

```
Test setup failed: critical/chats/test_public_chat_browsing.py:868: in prepare_devices
    self.community_2.join_community()
../views/chat_view.py:425: in join_community
    self.community_status_joined.wait_for_visibility_of_element(60)
../views/base_element.py:147: in wait_for_visibility_of_element
    raise TimeoutException(
 Device 2: Text by accessibility id:`status-tag-positive` is not found on the screen after wait_for_visibility_of_element
```

<br/><br/></td></tr></tbody></table><h4>Class TestCommunityMultipleDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3627169">test_community_message_send_check_timestamps_sender_username</a>, id: 702838 </b></td></tr><tr><td><p><blockquote><div>Device 2: Verifying that 'new message' is under today </div><div>Device 2: Looking for a message by text: new message </div></blockquote></p>

```
critical/chats/test_public_chat_browsing.py:364: in test_community_message_send_check_timestamps_sender_username
    self.channel_2.verify_message_is_under_today_text(new_message, self.errors, 60)
../views/chat_view.py:972: in verify_message_is_under_today_text
    message_element.wait_for_visibility_of_element(timeout)
../views/base_element.py:147: in wait_for_visibility_of_element
    raise TimeoutException(
 Device 2: ChatElementByText by xpath:`//*[starts-with(@text,'new message')]/ancestor::android.view.ViewGroup[@content-desc='chat-item']` is not found on the screen after wait_for_visibility_of_element
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/3eb3a132e76240f5a650b9f6d35de970?auth=a9a66755e0188bd6bf2d825e92b33a82#160">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/72a842ab59ca4c8bac9ff39bea765b9d?auth=4ab111537d89a7730bec2f60ac64a466#104">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestDeepLinksOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3627167">test_links_deep_links</a>, id: 702775 </b></td></tr><tr><td><p><blockquote><div>Device 1: Waiting 35s for dapp to load </div><div>Device 1: Find `Text` by `accessibility id`: `community-title` </div></blockquote></p>

```
critical/test_deep_and_universal_links.py:112: in test_links_deep_links
    if self.community_view.community_title.text != text:
../views/base_element.py:419: in text
    text = self.find_element().text
../views/base_element.py:79: in find_element
    raise NoSuchElementException(
 Device 1: Text by accessibility id: `community-title` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/da2e3b8ee7bb4a449380b852b16b4d97?auth=86586bae417f4c6361a78b65f9537fb0#126">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3627171">test_links_open_universal_links_from_chat</a>, id: 704613 </b></td></tr><tr><td><p><blockquote><div>Device 1: Looking for a message by text: https://status.app/c/ixyACjgKDVNOVCBjb21tdW5pdHkSHHJlcXVpcmUgMTAgU05UIEdvZXJsaSB0byB1c2UYASIHI2VhYjcwMAM=#zQ3shUeGnhM33QW4g9JfYfeLFAH9ZwbDboNYn5exCR7S3ii1y </div><div>Device 1: Find `Text` by `xpath`: `//android.view.ViewGroup[@content-desc='chat-item']//android.widget.TextView[contains(@text,'https://status.app/c/ixyACjgKDVNOVCBjb21tdW5pdHkSHHJlcXVpcmUgMTAgU05UIEdvZXJsaSB0byB1c2UYASIHI2VhYjcwMAM=#zQ3shUeGnhM33QW4g9JfYfeLFAH9ZwbDboNYn5exCR7S3ii1y')]` </div></blockquote></p>

```
critical/test_deep_and_universal_links.py:62: in test_links_open_universal_links_from_chat
    self.channel.chat_element_by_text(url).click_on_link_inside_message_body()
../views/chat_view.py:176: in click_on_link_inside_message_body
    self.message_body.click_inside_element_by_coordinate(rel_x=0.1, rel_y=0.9)
../views/base_element.py:357: in click_inside_element_by_coordinate
    location, size = self.get_element_coordinates()
../views/base_element.py:296: in get_element_coordinates
    size = element.size
../../../../status-app-prs-rerun@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py:275: in size
    size = self._execute(Command.GET_ELEMENT_RECT)["value"]
../../../../status-app-prs-rerun@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py:395: in _execute
    return self._parent.execute(command, params)
../../../../status-app-prs-rerun@tmp/venv/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:345: in execute
    self.error_handler.check_response(response)
../../../../status-app-prs-rerun@tmp/venv/lib/python3.10/site-packages/appium/webdriver/errorhandler.py:122: in check_response
    raise exception_class(msg=message, stacktrace=format_stacktrace(stacktrace))
 The element 'By.xpath: //android.view.ViewGroup[@content-desc='chat-item']//android.widget.TextView[contains(@text,'https://status.app/c/ixyACjgKDVNOVCBjb21tdW5pdHkSHHJlcXVpcmUgMTAgU05UIEdvZXJsaSB0byB1c2UYASIHI2VhYjcwMAM=#zQ3shUeGnhM33QW4g9JfYfeLFAH9ZwbDboNYn5exCR7S3ii1y')]' is not linked to the same object in DOM anymore; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#stale-element-reference-exception
E   Stacktrace:
E   io.appium.uiautomator2.common.exceptions.StaleElementReferenceException: The element 'By.xpath: //android.view.ViewGroup[@content-desc='chat-item']//android.widget.TextView[contains(@text,'https://status.app/c/ixyACjgKDVNOVCBjb21tdW5pdHkSHHJlcXVpcmUgMTAgU05UIEdvZXJsaSB0byB1c2UYASIHI2VhYjcwMAM=#zQ3shUeGnhM33QW4g9JfYfeLFAH9ZwbDboNYn5exCR7S3ii1y')]' is not linked to the same object in DOM anymore
E   	at io.appium.uiautomator2.model.ElementsCache.restore(ElementsCache.java:122)
E   	at io.appium.uiautomator2.model.ElementsCache.get(ElementsCache.java:153)
E   	at io.appium.uiautomator2.handler.GetRect.safeHandle(GetRect.java:40)
E   	at io.appium.uiautomator2.handler.request.SafeRequestHandler.handle(SafeRequestHandler.java:59)
E   	at io.appium.uiautomator2.server.AppiumServlet.handleRequest(AppiumServlet.java:277)
E   	at io.appium.uiautomator2.server.AppiumServlet.handleHttpRequest(AppiumServlet.java:271)
E   	at io.appium.uiautomator2.http.ServerHandler.channelRead(ServerHandler.java:68)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
E   	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
E   	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
E   	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
E   	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:435)
E   	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
E   	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:267)
E   	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:250)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
E   	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
E   	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:266)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
E   	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:345)
E   	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1294)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:366)
E   	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:352)
E   	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:911)
E   	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
E   	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:611)
E   	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:552)
E   	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:466)
E   	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:438)
E   	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:140)
E   	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
E   	at java.lang.Thread.run(Thread.java:1012)
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/da2e3b8ee7bb4a449380b852b16b4d97?auth=86586bae417f4c6361a78b65f9537fb0#50">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

--- a/src/status_im/common/signals/events.cljs
+++ b/src/status_im/common/signals/events.cljs
@@ -38,6 +38,12 @@
       "wallet"
       {:fx [[:dispatch [:wallet/signal-received event-js]]]}
 
+      "wallet.sign.transactions"
+      {:fx [[:dispatch
+             [:standard-auth/authorize-with-keycard
+              {:on-complete #(rf/dispatch [:keycard/sign-hash %
+                                           (first (transforms/js->clj event-js))])}]]]}
+
       "wallet.suggested.routes"
       {:fx [[:dispatch [:wallet/handle-suggested-routes (transforms/js->clj event-js)]]]}
 

--- a/src/status_im/common/standard_authentication/events.cljs
+++ b/src/status_im/common/standard_authentication/events.cljs
@@ -3,21 +3,35 @@
     [schema.core :as schema]
     [status-im.common.standard-authentication.enter-password.view :as enter-password]
     [status-im.common.standard-authentication.events-schema :as events-schema]
+    [status-im.contexts.keycard.pin.view :as keycard.pin]
     [taoensso.timbre :as log]
+    [utils.address]
     [utils.i18n :as i18n]
     [utils.re-frame :as rf]
     [utils.security.core :as security]))
 
+(rf/reg-fx :effects.keycard/call-on-auth-success
+ (fn [on-auth-success]
+   (when on-auth-success (on-auth-success ""))))
+
 (defn authorize
-  [{:keys [db]} [args]]
+  [{:keys [db]} [{:keys [on-auth-success keycard-supported?] :as args}]]
   (let [key-uid  (get-in db [:profile/profile :key-uid])
         keycard? (get-in db [:profile/profile :keycard-pairing])]
-    {:fx [[:effects.biometric/check-if-available
-           {:key-uid    key-uid
-            :on-success #(rf/dispatch [:standard-auth/authorize-with-biometric args])
-            :on-fail    (if keycard?
-                          #(rf/dispatch [:standard-auth/authorize-with-keycard args])
-                          #(rf/dispatch [:standard-auth/authorize-with-password args]))}]]}))
+    {:fx
+     [(if keycard?
+        (if keycard-supported?
+          [:effects.keycard/call-on-auth-success on-auth-success]
+          [:effects.utils/show-popup
+           {:title "This feature is not supported yet "
+            :content
+            "Keycard support is limited to logging in
+            and signing the sending transaction.
+            Use Status Desktop to access all functions."}])
+        [:effects.biometric/check-if-available
+         {:key-uid    key-uid
+          :on-success #(rf/dispatch [:standard-auth/authorize-with-biometric args])
+          :on-fail    #(rf/dispatch [:standard-auth/authorize-with-password args])}])]}))
 
 (schema/=> authorize events-schema/?authorize)
 (rf/reg-event-fx :standard-auth/authorize authorize)
@@ -94,6 +108,16 @@
         :button-icon-left    auth-button-icon-left
         :button-label        auth-button-label}])))
 
+(defn authorize-with-keycard
+  [_ [{:keys [on-complete]}]]
+  {:fx [[:dispatch
+         [:show-bottom-sheet
+          {:hide-on-background-press? false
+           :on-close                  #(rf/dispatch [:standard-auth/reset-login-password])
+           :content                   (fn []
+                                        [keycard.pin/auth {:on-complete on-complete}])}]]]})
+(rf/reg-event-fx :standard-auth/authorize-with-keycard authorize-with-keycard)
+
 (defn authorize-with-password
   [_ [{:keys [on-close theme blur?] :as args}]]
   {:fx [[:dispatch [:standard-auth/reset-login-password]]
@@ -110,7 +134,9 @@
 (rf/reg-event-fx
  :standard-auth/reset-login-password
  (fn [{:keys [db]}]
-   {:db (update db :profile/login dissoc :password :error)}))
+   {:db (-> db
+            (update :profile/login dissoc :password :error)
+            (update :keycard dissoc :pin))}))
 
 (rf/reg-fx
  :standard-auth/on-close

--- a/src/status_im/common/standard_authentication/standard_auth/slide_button/view.cljs
+++ b/src/status_im/common/standard_authentication/standard_auth/slide_button/view.cljs
@@ -8,7 +8,7 @@
 
 (defn view
   [{:keys [track-text customization-color auth-button-label on-auth-success on-auth-fail
-           auth-button-icon-left size blur? container-style disabled? dependencies]
+           auth-button-icon-left size blur? container-style disabled? dependencies keycard-supported?]
     :or   {container-style {:flex 1}}}]
   (let [theme           (quo.theme/use-theme)
         auth-method     (rf/sub [:auth-method])
@@ -21,6 +21,7 @@
                                           :auth-button-icon-left auth-button-icon-left
                                           :theme                 theme
                                           :blur?                 blur?
+                                          :keycard-supported?    keycard-supported?
                                           :biometric-auth?       biometric-auth?
                                           :on-auth-success       on-auth-success
                                           :on-auth-fail          on-auth-fail

--- a/src/status_im/contexts/keycard/effects.cljs
+++ b/src/status_im/contexts/keycard/effects.cljs
@@ -109,3 +109,22 @@
   [pairings]
   (keycard/set-pairings pairings))
 (rf/reg-fx :effects.keycard/set-pairing-to-keycard set-pairing-to-keycard)
+
+(defn sign
+  [{:keys [on-success on-failure] :as args}]
+  (log/debug "[keycard] sign")
+  (keycard/sign
+   (assoc
+    args
+    :on-success
+    (fn [response]
+      (log/debug "[keycard response succ] sign")
+      (when on-success
+        (on-success (transforms/js->clj response))))
+    :on-failure
+    (fn [response]
+      (log/warn "[keycard response fail] sign"
+                (error-object->map response))
+      (when on-failure
+        (on-failure (error-object->map response)))))))
+(rf/reg-fx :effects.keycard/sign sign)

--- a/src/status_im/contexts/keycard/events.cljs
+++ b/src/status_im/contexts/keycard/events.cljs
@@ -3,6 +3,8 @@
             status-im.contexts.keycard.login.events
             status-im.contexts.keycard.pin.events
             status-im.contexts.keycard.sheet.events
+            status-im.contexts.keycard.sign.events
+            [status-im.contexts.keycard.utils :as keycard.utils]
             [taoensso.timbre :as log]))
 
 (rf/reg-event-fx :keycard/on-check-nfc-enabled-success
@@ -55,3 +57,19 @@
  (fn [{:keys [db]} [{:keys [on-cancel-event-vector]}]]
    (log/debug "[keycard] nfc started success")
    {:db (assoc-in db [:keycard :on-nfc-cancelled-event-vector] on-cancel-event-vector)}))
+
+(rf/reg-event-fx :keycard/on-action-with-pin-error
+ (fn [{:keys [db]} [error]]
+   (log/debug "[keycard] get keys error: " error)
+   (let [tag-was-lost?     (keycard.utils/tag-lost? (:error error))
+         pin-retries-count (keycard.utils/pin-retries (:error error))]
+     (if tag-was-lost?
+       {:db (assoc-in db [:keycard :pin :status] nil)}
+       (if (nil? pin-retries-count)
+         {:effects.utils/show-popup {:title "wrong-keycard"}}
+         {:db (-> db
+                  (assoc-in [:keycard :application-info :pin-retry-counter] pin-retries-count)
+                  (update-in [:keycard :pin] assoc :status :error))
+          :fx [[:dispatch [:keycard/hide-connection-sheet]]
+               (when (zero? pin-retries-count)
+                 [:effects.utils/show-popup {:title "frozen-keycard"}])]})))))

--- a/src/status_im/contexts/keycard/login/events.cljs
+++ b/src/status_im/contexts/keycard/login/events.cljs
@@ -3,22 +3,6 @@
             [taoensso.timbre :as log]
             [utils.re-frame :as rf]))
 
-(rf/reg-event-fx :keycard.login/on-get-keys-error
- (fn [{:keys [db]} [error]]
-   (log/debug "[keycard] get keys error: " error)
-   (let [tag-was-lost?     (keycard.utils/tag-lost? (:error error))
-         pin-retries-count (keycard.utils/pin-retries (:error error))]
-     (if tag-was-lost?
-       {:db (assoc-in db [:keycard :pin :status] nil)}
-       (if (nil? pin-retries-count)
-         {:effects.utils/show-popup {:title "wrong-keycard"}}
-         {:db (-> db
-                  (assoc-in [:keycard :application-info :pin-retry-counter] pin-retries-count)
-                  (update-in [:keycard :pin] assoc :status :error))
-          :fx [[:dispatch [:keycard/hide-connection-sheet]]
-               (when (zero? pin-retries-count)
-                 [:effects.utils/show-popup {:title "frozen-keycard"}])]})))))
-
 (rf/reg-event-fx :keycard.login/on-get-keys-success
  (fn [{:keys [db]} [data]]
    (let [{:keys [key-uid encryption-public-key
@@ -56,28 +40,25 @@
                :key-uid             key-uid}]]}))))
 
 (rf/reg-event-fx :keycard.login/on-get-application-info-success
- (fn [{:keys [db]} [application-info]]
-   (let [profile (get-in db [:profile/profiles-overview (get-in db [:profile/login :key-uid])])
-         pin     (get-in db [:keycard :pin :text])
-         error   (keycard.utils/validate-application-info profile application-info)]
+ (fn [{:keys [db]} [application-info {:keys [key-uid on-read-fx]}]]
+   (let [error (keycard.utils/validate-application-info key-uid application-info)]
      (if error
        {:effects.utils/show-popup {:title (str error)}}
-       {:db                       (-> db
-                                      (assoc-in [:keycard :application-info] application-info)
-                                      (assoc-in [:keycard :pin :status] :verifying))
-        :effects.keycard/get-keys {:pin        pin
-                                   :on-success #(rf/dispatch [:keycard.login/on-get-keys-success %])
-                                   :on-failure #(rf/dispatch [:keycard.login/on-get-keys-error %])}}))))
+       {:db (-> db
+                (assoc-in [:keycard :application-info] application-info)
+                (assoc-in [:keycard :pin :status] :verifying))
+        :fx on-read-fx}))))
 
 (rf/reg-event-fx :keycard.login/cancel-reading-card
  (fn [{:keys [db]}]
    {:db (assoc-in db [:keycard :on-card-connected-event-vector] nil)}))
 
-(rf/reg-event-fx :keycard/read-card-and-login
- (fn [{:keys [db]}]
+(rf/reg-event-fx :keycard/read-card
+ (fn [{:keys [db]} [args]]
    (let [connected?   (get-in db [:keycard :card-connected?])
          event-vector [:keycard/get-application-info
-                       {:on-success #(rf/dispatch [:keycard.login/on-get-application-info-success %])}]]
+                       {:on-success #(rf/dispatch [:keycard.login/on-get-application-info-success %
+                                                   args])}]]
      (log/debug "[keycard] proceed-to-login")
      {:db (assoc-in db [:keycard :on-card-connected-event-vector] event-vector)
       :fx [[:dispatch

--- a/src/status_im/contexts/keycard/pin/events.cljs
+++ b/src/status_im/contexts/keycard/pin/events.cljs
@@ -9,13 +9,17 @@
                 (assoc-in [:keycard :pin :text] (.slice pin 0 -1))
                 (assoc-in [:keycard :pin :status] nil))}))))
 
+(rf/reg-fx :effects.keycard.pin/dispatch-on-complete
+ (fn [[on-complete new-pin]]
+   (on-complete new-pin)))
+
 (rf/reg-event-fx :keycard.pin/number-pressed
- (fn [{:keys [db]} [number max-numbers on-complete-event]]
+ (fn [{:keys [db]} [number max-numbers on-complete]]
    (let [pin     (get-in db [:keycard :pin :text])
          new-pin (str pin number)]
      (when (<= (count new-pin) max-numbers)
        {:db (-> db
                 (assoc-in [:keycard :pin :text] new-pin)
                 (assoc-in [:keycard :pin :status] nil))
-        :fx [(when (= (dec max-numbers) (count pin))
-               [:dispatch [on-complete-event]])]}))))
+        :fx [(when (and on-complete (= (dec max-numbers) (count pin)))
+               [:effects.keycard.pin/dispatch-on-complete [on-complete new-pin]])]}))))

--- a/src/status_im/contexts/keycard/pin/view.cljs
+++ b/src/status_im/contexts/keycard/pin/view.cljs
@@ -6,7 +6,7 @@
             [utils.re-frame :as rf]))
 
 (defn auth
-  [callback-event-key]
+  [{:keys [on-complete]}]
   (let [{:keys [text status]} (rf/sub [:keycard/pin])
         pin-retry-counter     (rf/sub [:keycard/pin-retry-counter])
         error?                (= status :error)]
@@ -23,4 +23,4 @@
       {:delete-key? true
        :on-delete   #(rf/dispatch [:keycard.pin/delete-pressed])
        :on-press    #(rf/dispatch [:keycard.pin/number-pressed % constants/pincode-length
-                                   callback-event-key])}]]))
+                                   on-complete])}]]))

--- a/src/status_im/contexts/keycard/sign/events.cljs
+++ b/src/status_im/contexts/keycard/sign/events.cljs
@@ -1,0 +1,29 @@
+(ns status-im.contexts.keycard.sign.events
+  (:require [utils.address]
+            [utils.re-frame :as rf]))
+
+(defn get-signature-map
+  [tx-hash signature]
+  {tx-hash {:r (subs signature 0 64)
+            :s (subs signature 64 128)
+            :v (str (js/parseInt (subs signature 128 130) 16))}})
+
+(rf/reg-event-fx :keycard/sign-hash
+ (fn [{:keys [db]} [pin-text tx-hash]]
+   (let [current-address (get-in db [:wallet :current-viewing-account-address])
+         path            (get-in db [:wallet :accounts current-address :path])
+         key-uid         (get-in db [:profile/profile :key-uid])]
+     {:fx [[:dispatch
+            [:keycard/read-card
+             {:key-uid    key-uid
+              :on-read-fx [[:effects.keycard/sign
+                            {:pin pin-text
+                             :path path
+                             :hash (utils.address/naked-address tx-hash)
+                             :on-success
+                             #(do
+                                (rf/dispatch [:keycard/hide-connection-sheet])
+                                (rf/dispatch
+                                 [:wallet/proceed-with-transactions-signatures
+                                  (get-signature-map tx-hash %)]))
+                             :on-failure #(rf/dispatch [:keycard/on-action-with-pin-error %])}]]}]]]})))

--- a/src/status_im/contexts/keycard/utils.cljs
+++ b/src/status_im/contexts/keycard/utils.cljs
@@ -16,8 +16,8 @@
    (re-matches #".*NFCError:100.*" error)))
 
 (defn validate-application-info
-  [profile {:keys [key-uid paired? pin-retry-counter puk-retry-counter] :as application-info}]
-  (let [profile-mismatch? (or (nil? profile) (not= (:key-uid profile) key-uid))]
+  [profile-key-uid {:keys [key-uid paired? pin-retry-counter puk-retry-counter] :as application-info}]
+  (let [profile-mismatch? (or (nil? profile-key-uid) (not= profile-key-uid key-uid))]
     (log/debug "[keycard] login-with-keycard"
                "empty application info" (empty? application-info)
                "no key-uid"             (empty? key-uid)

--- a/src/status_im/contexts/profile/profiles/view.cljs
+++ b/src/status_im/contexts/profile/profiles/view.cljs
@@ -239,7 +239,16 @@
         :profile-picture     profile-picture
         :card-style          style/login-profile-card}]
       (if keycard-pairing
-        [keycard.pin/auth :keycard/read-card-and-login]
+        [keycard.pin/auth
+         {:on-complete
+          (fn [pin-text]
+            (rf/dispatch
+             [:keycard/read-card
+              {:key-uid    key-uid
+               :on-read-fx [[:effects.keycard/get-keys
+                             {:pin        pin-text
+                              :on-success #(rf/dispatch [:keycard.login/on-get-keys-success %])
+                              :on-failure #(rf/dispatch [:keycard/on-action-with-pin-error %])}]]}]))}]
         [password-input])]
      (when-not keycard-pairing
        [quo/button

--- a/src/status_im/contexts/wallet/send/events.cljs
+++ b/src/status_im/contexts/wallet/send/events.cljs
@@ -698,8 +698,9 @@
      {:json-rpc/call [{:method     "wallet_createMultiTransaction"
                        :params     request-params
                        :on-success (fn [result]
-                                     (rf/dispatch [:wallet/add-authorized-transaction result])
-                                     (rf/dispatch [:hide-bottom-sheet]))
+                                     (when result
+                                       (rf/dispatch [:wallet/add-authorized-transaction result])
+                                       (rf/dispatch [:hide-bottom-sheet])))
                        :on-error   (fn [error]
                                      (log/error "failed to send transaction"
                                                 {:event  :wallet/send-transaction
@@ -709,6 +710,23 @@
                                                    {:id   :send-transaction-error
                                                     :type :negative
                                                     :text (:message error)}]))}]})))
+
+(rf/reg-event-fx :wallet/proceed-with-transactions-signatures
+ (fn [_ [signatures]]
+   {:json-rpc/call [{:method     "wallet_proceedWithTransactionsSignatures"
+                     :params     [signatures]
+                     :on-success (fn [result]
+                                   (when result
+                                     (rf/dispatch [:wallet/add-authorized-transaction result])
+                                     (rf/dispatch [:hide-bottom-sheet])))
+                     :on-error   (fn [error]
+                                   (log/error "failed to proceed-with-transactions-signatures"
+                                              {:event :wallet/proceed-with-transactions-signatures
+                                               :error error})
+                                   (rf/dispatch [:toasts/upsert
+                                                 {:id   :send-transaction-error
+                                                  :type :negative
+                                                  :text (:message error)}]))}]}))
 
 (rf/reg-event-fx
  :wallet/select-from-account

--- a/src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
+++ b/src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
@@ -257,7 +257,8 @@
                                         :transaction-type   transaction-type}]
                                       (when (and (not loading-suggested-routes?) route (seq route))
                                         [standard-auth/slide-button
-                                         {:size                :size-48
+                                         {:keycard-supported?  true
+                                          :size                :size-48
                                           :track-text          (if (= transaction-type :tx/bridge)
                                                                  (i18n/label :t/slide-to-bridge)
                                                                  (i18n/label :t/slide-to-send))
@@ -265,8 +266,7 @@
                                           :customization-color account-color
                                           :on-auth-success     #(rf/dispatch
                                                                  [:wallet/send-transaction
-                                                                  (security/safe-unmask-data
-                                                                   %)])
+                                                                  (security/safe-unmask-data %)])
                                           :auth-button-label   (i18n/label :t/confirm)}])]
            :gradient-cover?          true
            :customization-color      (:color account)}

